Requeue Service after Endpoint Deletion

swetharepakula opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
When the same service is deleted and quickly recreated, there is a race of whether the endpoints cache is up to date when processing the service creation causing the endpoints controller to wrongly see deleted endpoints as existing. This causes the endpoints to never be recreated.

This PR attempts to fix that race by adding a tracker similar to one in endpoints slice. The endpoints tracker will keep track of what the endpoints controller expects the state is. If any endpoint add, update, or deletion occurs that does not match the tracker's state, the associated service will be reprocessed.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #47711

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Endpoint controller requeues service after an endpoint deletion event occurs to confirm that deleted endpoints are undesired to mitigate the effects of an out of sync endpoint cache.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

swetharepakula commented:
/sig network

swetharepakula commented:
/sig apps

swetharepakula commented:
/remove-sig apps

kragniz commented:
/ok-to-test

aojea commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1990,6 +1990,200 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+	case <-blockNextAction:
+	}
+
+	// if endpoints cache has not updated becfore service update is registered
```
nit s/before/becfore/

aojea commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1990,6 +1990,200 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+	}
+	store.Add(ep1)
+	endpoints.onEndpointsUpdate(ep1)
+	if endpoints.queue.Len() != 1 {
```
can we have a race here if we assert before the queue is populated?

swetharepakula replied:
This test does not try to replicate the race, but just makes sure the service is added to the queue to be processed when an unexpected endpoints event comes through. `TestMultipleServiceChanges` covers the race condition.

aojea commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1990,6 +1990,200 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+		t.Errorf("Expected one service to be in the queue, found %d", endpoints.queue.Len())
+	}
+	// Clear the queue
+	for endpoints.queue.Len() != 0 {
```
we confirmed that the Len is 1 before, should be possible to have more than 1 at this moment?

swetharepakula replied:
That is true, this for loop is a remnant of an earlier iteration of the test. Updated so it just pops the one element.

swetharepakula commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -309,6 +319,38 @@ func (e *EndpointController) onServiceDelete(obj interface{}) {
+func (e *EndpointController) onEndpointsUpdate(obj interface{}) {
+	ep, ok := obj.(*v1.Endpoints)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Failed getting endpoint from %+v", obj))
```
Fixed in the update here and for the delete function.

swetharepakula commented on pkg/controller/endpoint/endpoints_tracker.go:
```diff
@@ -0,0 +1,75 @@
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
```
Thanks! The BUILD file should have all the right dependencies now.

swetharepakula commented on pkg/controller/endpoint/endpoints_tracker_test.go:
```diff
@@ -0,0 +1,240 @@
+
+func TestHandleEndpointTrackerUpdate(t *testing.T) {
+
+	tracker := newEndpointsTracker()
```
I initially envisioned the test where each case built upon the previous. Though for clarity I removed the duplicate test cases and a new tracker is initialized per test case. Now each case is independent of the previous.

swetharepakula commented on pkg/controller/endpoint/endpoints_tracker_test.go:
```diff
@@ -0,0 +1,240 @@
+		},
+		{
+			desc:           "check for different resource version, since now added",
+			updateEndpoint: ep2,
```
See above comment.

swetharepakula commented:
Thanks @aojea for the feedback and guidance! I have made updates based on your suggestions. Please let me know what you think.

swetharepakula commented:
/remove-sig apps

aojea commented:
heh, nice finding, this is somehow breaking leader election in client-go, we need to investigate why @swetharepakula, I pasted below some logs that may give us a hint

```
F0720 19:58:50.404098       8 controllermanager.go:293] leaderelection lost
```

```
2020-07-20T19:35:47.795032275Z stderr F E0720 19:35:47.794880       1 controller.go:152] Unable to remove old endpoints from kubernetes service: StorageError: key not found, Code: 1, Key: /registry/masterleases/172.18.0.3, ResourceVersion: 0, AdditionalErrorMsg:
```
@robscott you have a new TODO for the endpointslice migration (cc @liggitt ) :)

https://github.com/kubernetes/client-go/blob/a1d715839a207876281587dee364fb74f5cc153b/tools/leaderelection/resourcelock/endpointslock.go#L65

liggitt commented:
> 2020-07-20T19:35:47.795032275Z stderr F E0720 19:35:47.794880       1 controller.go:152] Unable to remove old endpoints from kubernetes service: StorageError: key not found, Code: 1, Key: /registry/masterleases/172.18.0.3, ResourceVersion: 0, AdditionalErrorMsg:


That's not leader election in client-go, that's the kubernetes master endpoints writer.

liggitt commented:
As noted in https://github.com/kubernetes/kubernetes/issues/85353#issuecomment-554575105, endpoints are 1:1 with their service matched by name. I don't think we should add tracking overhead for endpoints, and should just watch endpoints deletes and requeue the corresponding service name.

swetharepakula commented:
@aojea, that was very helpful! My understanding now is that the endpoints created by the endpointslock aren't in the tracker so the endpoint controller views it as an undesired endpoint and tries to delete them.

@liggitt Without tracking, are you thinking we should queue services only on delete or all endpoint events? Both situations seem like we would be enqueuing many services unnecessarily.

liggitt commented:
We've survived without tracking updates for ~5 years. I would start by doing a requeue on delete to fix the known issue. In most cases, the service will not exist and the sync will be very cheap.

swetharepakula commented:
In that case. I will update the PR with that solution and remove the tracker. Thanks for the suggestion @liggitt !

swetharepakula committed:
Requeue service after endpoint deletion

 - ensure endpoints that have been deleted and are desired are recreated
 despite a possibly out of date endpoint cache

swetharepakula commented:
/remove-sig apps

liggitt commented:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93030/pull-kubernetes-node-e2e/1285271293750415360 flake is #93210

swetharepakula commented:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93030/pull-kubernetes-e2e-gce-100-performance/1285371796182274048/ flake is #93217

aojea commented:
> I would start by doing a requeue on delete

let me understand it, the theory is
1. create service
2. delete service
3. create service
4. the endpoint object is deleted but still in cache, so the controller does not create new endpoints

in order to solve it, after 2. we receive a delete endpoint and we requeue it to sync again

with this approach, we may solve 4 but , if the endpoint delete arrives later, can this new sync delete the new endpoints created by 3?

liggitt commented:
> can this new sync delete the new endpoints created by 3?

the controller is the thing creating the endpoints. if the service is present in its cache at the beginning of its sync method, it will create the endpoints. if the service is absent from the cache at the beginning of the sync method it will delete the corresponding endpoints.

aojea commented:
thanks for the explanation :smile: 
/retest

aojea commented:
/lgtm
/assign @bowei

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+
+	endpoints.serviceStore.Add(svc)
+	endpoints.onServiceUpdate(svc)
+	timer := time.NewTimer(1 * time.Second)
+	select {
+	// blockNextAction should eventually unblock once server gets endpoint request.
+	case <-timer.C:
+		t.Errorf("Service Add should have caused a request to be sent to the test server")
+	case <-blockNextAction:
+	}
```
Would it be possible to move this logic into a helper function? It looks like it's repeated a few times.

swetharepakula replied:
Moved this logic into a helper function.

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+
+	endpoints.serviceStore.Add(svc)
+	endpoints.onServiceUpdate(svc)
+	timer := time.NewTimer(1 * time.Second)
```
Can we always guarantee that 1 second will be enough here?

swetharepakula replied:
Looking at the other tests, when waiting for the queue to be processed 1 second seemed to be used. But if there is a chance of increased flakiness, I can increase the timeout. What would you suggest?

robscott replied:
👍 That does seem reasonable. The other tests have not been flaky.

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
 	}
 }
 
+func TestMultipleServiceChanges(t *testing.T) {
```
There's a lot going on in this test, it might help to have a comment above the function explaining what this test is trying to cover and a high level overview of how it accomplishes that.

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+	}
+
+	var endpointsStore cache.Store
+	var endpoints *endpointController
```
```suggestion
	var controller *endpointController
```

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+			return
+		}
+		if endpoints == nil {
+
```
This feels like an unnecessary new line. On the other hand, a new line before the `if endpoints == nil` line would help match the rest of the structure here.

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+		if endpoints == nil {
+
+			res.WriteHeader(http.StatusInternalServerError)
+			res.Write([]byte("endpoints"))
```
This could use a more descriptive error message.

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+		t.Errorf("unexpected request: %v", req.RequestURI)
+		http.Error(res, "", http.StatusNotFound)
+	})
+	testServer := httptest.NewServer(mux)
```
Would it be possible to extract this and the handlerFunc into some kind of helper func?

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+	case <-blockNextAction:
+	}
+
+	// if endpoints cache has not updated before service update is registered
+	// Services add will not trigger a Create endpoint request.
```
Nit: The capitalization feels inconsistent here.

robscott commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+	ns := metav1.NamespaceDefault
+	testServer, _ := makeTestServer(t, ns)
+	defer testServer.Close()
+	endpoints := newController(testServer.URL, 0)
```
```suggestion
	controller := newController(testServer.URL, 0)
```

robscott reviewed (commented):
Thanks for the work on this! The actual change to the controller LGTM, just had some suggestions and questions around the tests.

robscott commented:
Thanks for the work on this!

/lgtm
/assign @freehan

swetharepakula commented:
/retest

robscott commented:
/cc @MrHohn

robscott commented:
/priority important-soon

MrHohn reviewed (approved):
Thanks for the fix and nice test!
/approve

BenTheElder commented:
/milestone v1.19

robscott commented:
/retest
